test/kad-dht.js

```diff
@@ -1,112 +1,692 @@
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => jsDf.spawn({ initOptions: { bits: 512 } }, cb)
-      ], (err, nodes) => {
+const jsDf = DaemonFactory.create({ type: 'proc', exec: IPFS })
+
+const getConfig = (bootstrap) => ({
+  Bootstrap: bootstrap,
+  Discovery: {
+    MDNS: {
+      Enabled: false
+    },
+    webRTCStar: {
+      Enabled: false
+    }
+  }
+})
+
+const spawnGoDaemon = (bootstrap = [], callback) => {
+  goDf.spawn({
+    initOptions: { bits: 1024 },
+    config: getConfig(bootstrap)
+  }, callback)
+}
+
+const spawnJsDaemon = (bootstrap = [], callback) => {
+  jsDf.spawn({
+    initOptions: { bits: 512 },
+    config: getConfig(bootstrap),
+    libp2p: {
+      config: {
+        dht: {
+          enabled: true
+        }
+      }
+    }
+  }, callback)
+}
+
+const getBootstrapAddr = (node, callback) => {
+  node.api.id((err, res) => {
+    expect(err).to.not.exist()
+    expect(res.id).to.exist()
+
+    callback(null, res.addresses[0])
+  })
+}
+
+const addFileAndCat = (addDaemon, catDaemons, data, callback) => {
+  addDaemon.api.add(data, (err, res) => {
+    expect(err).to.not.exist()
+
+    each(catDaemons, (daemon, cb) => {
+      daemon.api.cat(res[0].hash, (err, res) => {
         expect(err).to.not.exist()
-        goD1 = nodes[0]
-        goD2 = nodes[1]
-        goD3 = nodes[2]
-        jsD = nodes[3]
-        done()
+        expect(res).to.equalBytes(data)
+        cb()
+      })
+    }, callback)
+  })
+}
+
+describe('kad-dht', () => {
+  describe('kad-dht with a single bootstrap node', () => {
+    describe('a JS network', () => {
+      let bootstrapAddr
+      const nodes = []
+      const data = crypto.randomBytes(9001)
+
+      // spawn bootstrap daemon and get address
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        spawnJsDaemon([], (err, node) => {
+          expect(err).to.not.exist()
+          nodes.push(node)
+
+          getBootstrapAddr(nodes[0], (err, res) => {
+            expect(err).to.not.exist()
+            bootstrapAddr = res
+            done()
+          })
+        })
+      })
+
+      // spawn daemons
+      before(function (done) {
+        this.timeout(70 * 1000)
+        const peersToSpawn = 3
+        const bootstrap = bootstrapAddr ? [bootstrapAddr] : []
+
+        times(peersToSpawn, (_, next) => {
+          spawnJsDaemon(bootstrap, (err, node) => {
+            expect(err).to.not.exist()
+            nodes.push(node)
+            next(err, node)
+          })
+        }, (err) => {
+          expect(err).to.not.exist()
+
+          const testDhtReady = (cb) => {
+            nodes[0].api.swarm.peers((err, peers) => {
+              if (err) {
+                return cb(err)
+              }
+              cb(null, peers.length === peersToSpawn)
+            })
+          }
```

Comment:
This function seems to only be testing if the first node is connected to all the other peers. It means that either:

- The name _testDhtReady_ is incorrect
- It is creating a full mesh which will then defeat a bit of the purpose of testing the DHT